lang/cs/Org.Apache.REEF.IMRU/OnREEF/StateMachine/SystemStateMachine.cs

```diff
@@ -0,0 +1,102 @@
+    /// </summary>
+    internal sealed class SystemStateMachine
+    {
+        internal readonly static IDictionary<StateTransition<SystemState, SystemStateEvent>, SystemState> Transitions = 
+            new ReadOnlyDictionary<StateTransition<SystemState, SystemStateEvent>, SystemState>(
+                new Dictionary<StateTransition<SystemState, SystemStateEvent>, SystemState>()
+        {
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.WaitingForEvaluator, SystemStateEvent.AllContextsAreReady), SystemState.SubmittingTasks },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.WaitingForEvaluator, SystemStateEvent.NotRecoverable), SystemState.Fail },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.SubmittingTasks, SystemStateEvent.AllTasksAreRunning), SystemState.TasksRunning },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.SubmittingTasks, SystemStateEvent.FailedEvaluator), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.SubmittingTasks, SystemStateEvent.FailedTask), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.TasksRunning, SystemStateEvent.FailedEvaluator), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.TasksRunning, SystemStateEvent.FailedTask), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.TasksRunning, SystemStateEvent.AllTasksAreCompleted), SystemState.TasksCompleted },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.ShuttingDown, SystemStateEvent.Recover), SystemState.WaitingForEvaluator },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.ShuttingDown, SystemStateEvent.NotRecoverable), SystemState.Fail }
+        });
+
+        private readonly object _stateLock = new object();
+        private SystemState _currentState;
+
+        /// <summary>
+        /// create a new SysteState starting from WaitingForEvaluator
+        /// </summary>
+        internal SystemStateMachine()
+        {
+            _currentState = SystemState.WaitingForEvaluator;
+        }
+
+        /// <summary>
+        /// Returns current state
+        /// </summary>
+        internal SystemState CurrentState
+        {
+            get
+            {
+                lock (_stateLock)
+                {
+                    return _currentState;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Based on the event and current state to determine the next state
+        /// </summary>
+        /// <param name="systemEvent"></param>
+        /// <returns></returns>
+        internal SystemState GetNext(SystemStateEvent systemEvent)
```

Comment:
Yes, but if someone calls this, and then takes action based on it which results in a call to `MoveNext`, we have a potential race condition on our hand. If there is no caller to this method right now, I'm in favor of removing it in the name of API simplicity. We can always add similar functionality later if we need it.